sql-plugin/src/main/scala/com/nvidia/spark/rapids/rowBasedExpressions.scala

```diff
@@ -0,0 +1,145 @@
+
+  def rowExpression(children: Seq[Expression]): SparkExpr
+
+  def nullSafe: Boolean
+
+  private def prepareEvaluation(expression: Expression): Expression = {
+    val serializer = new JavaSerializer(new SparkConf()).newInstance
+    val resolver = ResolveTimeZone
+    val expr = resolver.resolveTimeZones(expression)
+    serializer.deserialize(serializer.serialize(expr))
+  }
+
+  private def evaluateWithoutCodegen(
+    expression: Expression, inputRow: InternalRow = EmptyRow): Any = {
+    expression.foreach {
+      case n: Nondeterministic => n.initialize(0)
+      case _ =>
+    }
+    expression.eval(inputRow)
+  }
+
+  private def evalRow(children: Seq[Expression], inputRow: InternalRow): Any = {
+    def expr = prepareEvaluation(rowExpression(children))
+    evaluateWithoutCodegen(expr, inputRow)
+  }
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    val cpuExprStart = System.nanoTime
+    val prepareArgsEnd = System.nanoTime
+
+    val childTypes = children.map(_.dataType)
+    // These child columns will be closed by `ColumnarToRowIterator`.
+    // 1 Evaluate all the children GpuExpressions to columns
+    val argCols = children.safeMap(GpuExpressionsUtils.columnarEvalToColumn(_, batch))
+    try {
+      // 2 Convert the argument columns to row.
+      val retConverter = GpuRowToColumnConverter.getConverterForType(dataType, nullable)
+      val retType = GpuColumnVector.convertFrom(dataType, nullable)
+      val retRow = new GenericInternalRow(size = 1)
+      closeOnExcept(new HostColumnVector.ColumnBuilder(retType, batch.numRows)) { builder =>
+        new ColumnarToRowIterator(
+            Iterator.single(new ColumnarBatch(argCols.toArray, batch.numRows())),
+            NoopMetric,
+            NoopMetric,
+            NoopMetric,
+            NoopMetric,
+            nullSafe).foreach { row =>
+          // 3 Row by row, convert the argument row to literal expressions to pass to the 
+          //   CPU Spark expression, then evaluate the CPU Spark expression and cache the 
+          //   result row.
+          val newChildren = childTypes.zip(row.toSeq(childTypes)).map { case (dt, value) =>
+            Literal.create(value, dt)
+          }
+          retRow.update(0, evalRow(newChildren, row))
+          retConverter.append(retRow, 0, builder)
+        }
+        // 4 Build a result column from the cache of result rows
+        closeOnExcept(builder.buildAndPutOnDevice()) { resultCol =>
+          val cpuRunningTime = System.nanoTime - prepareArgsEnd
+          // Use log to record the eclipsed time for the Expression running before
+          // figuring out how to support Spark metrics in this expression.
+          logDebug(s"It took ${cpuRunningTime} ns to run Expression ${prettyName}, " +
+            s"and ${prepareArgsEnd - cpuExprStart} ns to get the input from children.")
+          GpuColumnVector.from(resultCol, dataType)
+        }
+      }
+    } catch {
+      case e: Exception =>
```

Comment:
Why catch to just rethrow?

Reply:
This was supposed to wrap the underlying exception in SparkException, but was not fully implemented, adding that now.